Multiplatform locking

Xorboo opened the pull request:
In regards to [issue #14](https://github.com/trbs/pid/issues/14)
1. Using [`msvcrt.locking`](https://docs.python.org/2/library/msvcrt.html#msvcrt.locking) to lock a file on windows (same call is used in [fasteners](https://github.com/harlowja/fasteners/blob/master/fasteners/process_lock.py#L237)). However, it requires some logic changes, because reading/writing to a locked file from a second handler seems to be impossible on windows and throws `PermissionError[13]` (while opening still works). I had to:
    - Disable some tests that try to read/write to locked file directly. This shouldn't be used in a normal program anyway.
    - Raise a new windows-only exception on `allow_samepid` flag, as otherwise second pidfile will try to read a file and raise an exception.
    - Manually try to read from a file when opening it to check if file is locked.

2. Using [psutil](https://pypi.org/project/psutil/) for checking if process is running. `os.kill()` doesn't work on widows, so we have to change it as well. Again, [same](https://github.com/giampaolo/psutil/blob/master/psutil/_psutil_posix.c#L77) call on POSIX, [custom](https://github.com/giampaolo/psutil/blob/master/psutil/arch/windows/process_info.c#L359) one on windows

3. Ignoring some commands on windows, as they are only available on Unix:
    - `os.fchmod()`
    - `os.fchown()`
    - `os.getgid()`

All tests run smoothly and decorator seems to be working fine in my project.

*Notes:*
 - I'm not sure if I changed `setup.py` correctly, needs checking.
 - Running two scripts at exactly the same time doesn't stop any of them, is there is the same racing condition on Linux?
 - `idna` (required by `coverage`) does not support python2.6 anymore, so i removed 2.6 support
 - `psutil` does not support python3.3, removed it as well.
 - Is there a way to somehow cover windows-only part of the code with tests?

trbs commented on pid/__init__.py:
```diff
@@ -162,7 +158,7 @@ def create(self):
         if self.lock_pidfile:
             try:
-                fcntl.flock(self.fh.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
+                fasteners.lock.try_lock(self.fh)
```
https://github.com/harlowja/fasteners/blob/master/fasteners/lock.py#L257

`try_lock` expects a python lock not a file handler ?

Xorboo commented:
Hey there, took me a while to get my hands on it:

1. I added windows platform to Travis, python is not officially supported, but we can [install it](https://github.com/WeblateOrg/translation-finder/blob/master/.travis.yml#L20-L33) via shell commands, seems to be working well! Except for python3.4, for which psutil seems to require `Microsoft Visual C++ 10.0` for some reason, resulting in a fail build.

2. Also noticed a slight difference in Python2 and Python3. When trying to read a locked pidfile - linux version successfully gets pid value, buton on windows python2 gets and empty string, while python3 raises an `Access denied` exception. Only matters for one test case.

3. Test coverage seems to be on the same 95% after including most of the python versions on windows platform, so i could remove all windows-related branch excludes.

4. And now regarding psutil. I can try copying their implementation, although it is mostly c++ code with wrappers (and i haven't really done that before) and it will complicate this module by quite a lot. Also I'm not sure, how should we properly mention psutil and their BSD license in that case?

trbs commented:
I don't think it's a big deal not supporting `Python 3.4` on Windows.

trbs commented:
I really appreciate all your work on this !
It's a large patch so it might take me some time to have the time to dive into this and merge it.

I probably also want to migrate to `pytest` before merging this.
Such that we can use things like this for the tests:
```
@pytest.mark.skipif(sys.platform == 'win32',
                    reason="does not run on windows")
```

spumer commented:
Any updates?

trbs commented:
I'm still hesitant to merge this as is.
Let's try and break this up into multiple smaller PR's to make it more manageable.

I think we should be able to make the `psutil` library optional, using it if available and requiring it for windows support.

(If a company is reading this and wants to sponsor some of the work that would be much appreciated)

Xorboo commented:
@trbs, Another long evening and done!

- `psutils` now only used on windows
- removed `xenial` requirement (it seems to be working fine without it now)
- added CI for 3.8
- some additional changes to tests, like using `tempfile` instead of hardcoded `/tmp` dir

One thing I haven't thought about is what should we do with `DEFAULT_PID_DIR = "/var/run/"` in case of Windows, I'm not really sure what folder if suitable here, any ideas?

trbs commented:
Please check with the latest master version.

Since these where quite large changes, I hope everybody can start testing this on all platforms.

Thanks for all the help on having PidFile gain win32 support ! 🥇

Xorboo commented:
Ah, pity that the PR has been split instead of merging a PR, so I don't get to have a personal commit in the repo. But I see you've made a lot of adjustments and improvements, so it doesn't matter as much, as long as we have Windows support now 😊 Thanks for taking the time, @trbs

trbs commented:
I'm more then happy to give you recognition ! 
What about creating a PR and update the AUTHORS list ?

Xorboo commented:
Thanks, glad to do it, check #30 

PS Also can confirm that pidfile seems to be working just fine on windows in 3.0
